File: primeNumbers/primes.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

// Parse string input into a numeric limit, supporting exponential notation (e.g., "10**5")
func parseLimit(input string) int64 {
	idx := strings.Index(input, "**")
	if idx != -1 {
		base, err1 := strconv.ParseInt(input[:idx], 10, 64)
		exp, err2 := strconv.ParseInt(input[idx+2:], 10, 64)
		if err1 == nil && err2 == nil {
			res := int64(1)
			for i := int64(0); i < exp; i++ {
				res *= base
			}
			return res
		}
	}
	val, _ := strconv.ParseInt(input, 10, 64)
	return val
}

// Method 1: Sieve of Eratosthenes
func sieveEratosthenes(n int64) (int64, int64) {
	if n < 2 {
		return 0, 0
	}

	maxK := (n - 3) / 2
	marked := make([]bool, maxK+1)

	count := int64(1)
	last := int64(2)

	for p := int64(3); p*p <= n; p += 2 {
		k := (p - 3) / 2
		if !marked[k] {
			for j := p * p; j <= n; j += 2*p {
				marked[(j-3)/2] = true
			}
		}
	}

	for k := int64(0); k <= maxK; k++ {
		if !marked[k] {
			count++
			last = 2*k + 3
		}
	}

	return count, last
}

// Method 2: Naive Trial Division
func naiveTrialDivision(n int64) (int64, int64) {
	if n < 2 {
		return 0, 0
	}

	count := int64(1)
	last := int64(2)

	for num := int64(3); num <= n; num += 2 {
		isPrime := true
		for i := int64(3); i*i <= num; i += 2 {
			if num%i == 0 {
				isPrime = false
				break
			}
		}
		if isPrime {
			count++
			last = num
		}
	}

	return count, last
}

func main() {
	input := "100"
	if len(os.Args) > 1 {
		input = os.Args[1]
	}

	n := parseLimit(input)
	if n <= 0 {
		fmt.Fprintln(os.Stderr, "Error: Invalid limit")
		os.Exit(1)
	}

	fmt.Printf("Target: primes in range [0, %d]\n", n)
	fmt.Println("Benchmarking methods sequentially...")
	fmt.Println("---")

	// Benchmark Naive Trial Division
	startNaive := time.Now()
	naiveCount, _ := naiveTrialDivision(n) // Fixed: Using '_' to ignore unused return value
	naiveTime := time.Since(startNaive).Seconds()

	// Benchmark Sieve of Eratosthenes
	startSieve := time.Now()
	sieveCount, sieveLast := sieveEratosthenes(n)
	sieveTime := time.Since(startSieve).Seconds()

	// Verify algorithms yield identical results
	if naiveCount != sieveCount {
		fmt.Fprintf(os.Stderr, "ERROR: Result mismatch. Naive: %d, Sieve: %d\n", naiveCount, sieveCount)
		os.Exit(1)
	}

	fmt.Printf("Found: %d primes | Last prime: %d\n\n", sieveCount, sieveLast)
	fmt.Println("Execution Times:")
	fmt.Printf("  %-32s : %.6fs\n", "Naive Trial Division (Go)", naiveTime)
	fmt.Printf("  %-32s : %.6fs\n", "Sieve of Eratosthenes (Go)", sieveTime)
	fmt.Println("---")
}

File: primeNumbers/primes.rs
use std::env;
use std::time::Instant;

// Parse string input into a numeric limit, supporting exponential notation (e.g., "10**5")
fn parse_limit(input: &str) -> Option<u64> {
    if let Some(idx) = input.find("**") {
        let base: u64 = input[..idx].parse().ok()?;
        let exp: u32 = input[idx + 2..].parse().ok()?;
        Some(base.pow(exp))
    } else {
        input.parse().ok()
    }
}

// Method 1: Sieve of Eratosthenes
fn sieve_eratosthenes(n: u64) -> (u64, u64) {
    if n < 2 { return (0, 0); }

    let max_k = ((n - 3) / 2) as usize;
    let mut marked = vec![false; max_k + 1];

    let mut count = 1;
    let mut last = 2;

    let mut p: u64 = 3;
    while p * p <= n {
        let k = ((p - 3) / 2) as usize;
        if !marked[k] {
            let mut j = p * p;
            while j <= n {
                marked[((j - 3) / 2) as usize] = true;
                j += 2 * p;
            }
        }
        p += 2;
    }

    for k in 0..=max_k {
        if !marked[k] {
            count += 1;
            last = 2 * (k as u64) + 3;
        }
    }

    (count, last)
}

// Method 2: Naive Trial Division
fn naive_trial_division(n: u64) -> (u64, u64) {
    if n < 2 { return (0, 0); }

    let mut count = 1;
    let mut last = 2;

    let mut num = 3;
    while num <= n {
        let mut is_prime = true;
        let mut i = 3;
        while i * i <= num {
            if num % i == 0 {
                is_prime = false;
                break;
            }
            i += 2;
        }
        if is_prime {
            count += 1;
            last = num;
        }
        num += 2;
    }

    (count, last)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input = if args.len() > 1 { &args[1] } else { "100" };

    let n = match parse_limit(input) {
        Some(val) => val,
        None => {
            eprintln!("Error: Invalid limit");
            std::process::exit(1);
        }
    };

    println!("Target: primes in range [0, {}]", n);
    println!("Benchmarking methods sequentially...");
    println!("---");

    // Benchmark Naive Trial Division
    let start_naive = Instant::now();
    let (naive_count, _) = naive_trial_division(n); // Fixed: Using '_' to suppress warning
    let naive_time = start_naive.elapsed().as_secs_f64();

    // Benchmark Sieve of Eratosthenes
    let start_sieve = Instant::now();
    let (sieve_count, sieve_last) = sieve_eratosthenes(n);
    let sieve_time = start_sieve.elapsed().as_secs_f64();

    // Verify algorithms yield identical results
    if naive_count != sieve_count {
        eprintln!("ERROR: Result mismatch. Naive: {}, Sieve: {}", naive_count, sieve_count);
        std::process::exit(1);
    }

    println!("Found: {} primes | Last prime: {}\n", sieve_count, sieve_last);
    println!("Execution Times:");
    println!("  {:<32} : {:.6}s", "Naive Trial Division (Rust)", naive_time);
    println!("  {:<32} : {:.6}s", "Sieve of Eratosthenes (Rust)", sieve_time);
    println!("---");
}
